src/runtime/converterextensions.cs

```diff
@@ -0,0 +1,185 @@
+            if (type == null) throw new ArgumentNullException(nameof(type));
+
+            foreach (var encoder in clrToPython.GetOrAdd(type, GetEncoders))
+            {
+                var result = encoder.TryEncode(obj);
+                if (result != null) return result;
+            }
+
+            return null;
+        }
+
+        static readonly ConcurrentDictionary<Type, IPyObjectEncoder[]>
+            clrToPython = new ConcurrentDictionary<Type, IPyObjectEncoder[]>();
+        static IPyObjectEncoder[] GetEncoders(Type type)
+        {
+            lock (encoders)
+            {
+                return encoders.Where(encoder => encoder.CanEncode(type)).ToArray();
+            }
+        }
+        #endregion
+
+        #region Decoding
+        static readonly ConcurrentDictionary<TypePair, Converter.TryConvertFromPythonDelegate>
+            pythonToClr = new ConcurrentDictionary<TypePair, Converter.TryConvertFromPythonDelegate>();
+        internal static bool TryDecode(IntPtr pyHandle, IntPtr pyType, Type targetType, out object result)
+        {
+            if (pyHandle == IntPtr.Zero) throw new ArgumentNullException(nameof(pyHandle));
+            if (pyType == IntPtr.Zero) throw new ArgumentNullException(nameof(pyType));
+            if (targetType == null) throw new ArgumentNullException(nameof(targetType));
+
+            var decoder = pythonToClr.GetOrAdd(new TypePair(pyType, targetType), pair => GetDecoder(pair.PyType, pair.ClrType));
+            result = null;
+            if (decoder == null) return false;
+            return decoder.Invoke(pyHandle, out result);
+        }
+
+        static Converter.TryConvertFromPythonDelegate GetDecoder(IntPtr sourceType, Type targetType)
+        {
+            IPyObjectDecoder decoder;
+            using (var pyType = new PyObject(sourceType))
+            {
+                lock (decoders)
+                {
+                    decoder = decoders.Find(d => d.CanDecode(pyType, targetType));
+                    if (decoder == null) return null;
+                }
+            }
+
+            var decode = genericDecode.MakeGenericMethod(targetType);
+
+            bool TryDecode(IntPtr pyHandle, out object result)
+            {
+                using (var pyObj = new PyObject(Runtime.SelfIncRef(pyHandle)))
+                {
+                    var @params = new object[] { pyObj, null };
+                    bool success = (bool)decode.Invoke(decoder, @params);
+                    result = @params[1];
+                    return success;
+                }
+            }
+
+            return TryDecode;
+        }
+
+        static readonly MethodInfo genericDecode = typeof(IPyObjectDecoder).GetMethod(nameof(IPyObjectDecoder.TryDecode));
+
+        #endregion
+
+        internal static void Reset()
+        {
+            lock (encoders)
+                lock (decoders)
+                {
+                    clrToPython.Clear();
```

Comment:
why do we need clrToPython and pythonToClr if we already have decoders and encoders?

Reply:
Got it, a single instance of an encoder/decoder can handle many types so these containers lazily process the list of encoders and decoders on each type encountered.